### PR TITLE
Add prop tests

### PR DESCRIPTION
**Pull request by buffalojoec**
#### Problem
The funnel program has some extensive test coverage, but it's missing
regression testing to ensure the arithmetic in calculating distributions
is tight.

#### Summary of Changes
Add prop tests to test the distribution arithmetic.

### CODE REVIEW

File: program/src/processor.rs [hunk 1]
```diff
@@ -119,3 +119,75 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
...
+            }
+
+            // Step 2.
+            let holders_reward = amount.checked_div(2);
+            if holders_reward.is_none() {
+                // If the holders rewards division fails, the function should
+                // return an error.
+                prop_assert_eq!(result, Err(ProgramError::InvalidInstructionData));
+                return Ok(());
+            }
```
Comment (joncinque) [inline program/src/processor.rs:166]: This should be unwrappable

File: program/src/processor.rs [hunk 2]
```diff
@@ -119,3 +119,75 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
...
+            let stakers_reward = stakers_reward_num.unwrap().checked_div(5);
+            if stakers_reward.is_none() {
+                // If the stakers rewards denominator division fails, the
+                // function should return an error.
+                prop_assert_eq!(result, Err(ProgramError::InvalidInstructionData));
+                return Ok(());
+            }
```
Comment (joncinque) [inline program/src/processor.rs:157]: If we set the limit for `amount` to `u64 / 2`, this should be unwrappable

File: program/src/processor.rs [hunk 3]
```diff
@@ -119,3 +119,75 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
...
+            let treasury_reward = incr_treasury_reward.unwrap().checked_sub(holders_reward.unwrap());
+            if treasury_reward.is_none() {
+                // If the holders rewards subtraction fails, the function should
+                // return an error.
+                prop_assert_eq!(result, Err(ProgramError::InvalidInstructionData));
+                return Ok(());
+            } else {
```
Comment (joncinque) [inline program/src/processor.rs:182]: This should also be unwrappable

**Review by joncinque — COMMENTED** [review]
Just a few comments to make this testing even more useful

File: program/src/processor.rs [hunk 4]
```diff
@@ -138,3 +138,65 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
...
+            // 3. treasury rewards
+
+            // Step 1.
+            let stakers_reward_num = amount.checked_mul(2);
+            if stakers_reward_num.is_none() {
+                // If the stakers rewards numerator multiplication breaks the
+                // u64 ceiling, the function should return an error.
+                prop_assert_eq!(result, Err(ProgramError::InvalidInstructionData));
+                return Ok(());
+            }
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/src/processor.rs]: Oops, my bad. Forgot to cut that one out, too.

**Review by joncinque — APPROVED** [review]

**Pull request merged**